src/lua/LuaFileSystem.cpp

```diff
@@ -7,15 +7,107 @@
 #include "LuaObject.h"
 #include "Pi.h"
 
-/*
- * Interface: FileSystem
- *
- * A global table that provides access to the filesystem.
- *
- * This interface is protected. It can only be used by scripts in Pioneer's
- * data directory. Mods and scripts in the user directory that try to use it
- * will get a Lua error.
- */
+#include "core/StringUtils.h"
+
+struct SplitURI
+{
+	std::string		 full_path;
+	std::string_view scheme;
+	std::string_view rel_path;
+};
+
+SplitURI ParseURIArg(lua_State* l, int numArg)
+{
+	SplitURI res;
+	size_t path_size;
+	const char* path_str = luaL_checklstring(l, numArg, &path_size);
+
+	res.full_path = std::string(path_str, path_size);
+	size_t sep_pos = res.full_path.find_first_of("://");
+	if (sep_pos != std::string::npos)
+	{
+		res.scheme = std::string_view(path_str, sep_pos);
+		sep_pos += 3;
+		res.rel_path = std::string_view(path_str + sep_pos, path_size - sep_pos);
+	}
+	return res;
+}
+
+struct FileSource
+{
+	FileSystem::FileSource* source;
+	bool					is_read_write;
+};
+
+FileSource DetermineFileSource(std::string_view scheme)
+{
+	FileSource res;
+	if (compare_ci(scheme, "user"))
+	{
+		res.source = &FileSystem::userFiles;
+		res.is_read_write = true;
+	} else if (compare_ci(scheme, "data"))
+	{
+		res.source = &FileSystem::gameDataFiles;
+		res.is_read_write = false;
+	} else
+	{
+		res.source = nullptr;
+		res.is_read_write = false;
+	}
+	return res;
+}
+
+static lua_CFunction l_original_io_open = nullptr;
+
+void LuaFileSystem::register_raw_io_open_function(lua_CFunction open)
+{
+	l_original_io_open = open;
+}
+
+int LuaFileSystem::l_patched_io_open(lua_State* L)
+{
+	SplitURI split_uri = ParseURIArg(L, 1);
+	if (split_uri.scheme.empty())
+	{
+		return luaL_error(L, "'%s' does not have a valid scheme, must be user:// or data://", split_uri.full_path.c_str());
+	}
+
+	const FileSource fs = DetermineFileSource(split_uri.scheme);
+	if (nullptr == fs.source)
+	{
+		return luaL_error(L, "'%s' does not have a valid scheme, must be user:// or data://", split_uri.full_path.c_str());
+	}
```

Comment:
This block (due to the way DetermineFileSource is implemented) can be condensed to:

```suggestion
	SplitURI split_uri = ParseURIArg(L, 1);
	const FileSource fs = DetermineFileSource(split_uri.scheme);
	if (nullptr == fs.source)
	{
		return luaL_error(L, "Path '%s' does not contain a valid scheme, must be user:// or data://", split_uri.full_path.c_str());
	}
```

This is arguably better, especially if a `data://` scheme is implied (e.g. for implementing custom data assets in lua) - unsure if an implied scheme is needed for the first implementation however.